AttributeError: ReprFailDoctest instance has no attribute 'reprcrash' with --doctest-modules, --junit-xml and at a failing doctest.

Originally reported by: **Richard Barrell (BitBucket: [rbarrell](http://bitbucket.org/rbarrell), GitHub: [rbarrell](http://github.com/rbarrell))**

----------------------------------------

Reproduction:

in `a/__init__.py`:

```
#!python

def foo():
    """
    >>> 1 + 1
    3
    """
    pass
```

Without --junit-xml:

```
#!text

$ py.test a --doctest-modules
============================= test session starts ==============================
platform linux2 -- Python 2.7.5 -- py-1.4.26 -- pytest-2.7.0
rootdir: /home/RichardB/a, inifile: 
plugins: cov
collected 1 items

a/__init__.py F

=================================== FAILURES ===================================
_______________________________ [doctest] a.foo ________________________________
002     """
003     >>> 1 + 1
Expected:
    3
Got:
    2

/home/RichardB/a/__init__.py:3: DocTestFailure
=========================== 1 failed in 0.01 seconds ===========================
```

And that's fine. However, when I try with --junit-xml:

```
#!text

$ py.test a --doctest-modules --junit-xml=junit.xml
============================= test session starts ==============================
platform linux2 -- Python 2.7.5 -- py-1.4.26 -- pytest-2.7.0
rootdir: /home/RichardB/a, inifile: 
plugins: cov
collected 1 items

a/__init__.py F
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/main.py", line 84, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/main.py", line 122, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/main.py", line 142, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 393, in execute
INTERNALERROR>     return wrapped_call(method(*args), self.execute)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 113, in wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 138, in get_result
INTERNALERROR>     py.builtin._reraise(*ex)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 123, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/runner.py", line 65, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/runner.py", line 75, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/runner.py", line 123, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/junitxml.py", line 193, in pytest_runtest_logreport
INTERNALERROR>     self.append_failure(report)
INTERNALERROR>   File "/home/RichardB/buildouts/pytest/eggs/pytest-2.7.0-py2.7.egg/_pytest/junitxml.py", line 132, in append_failure
INTERNALERROR>     message = report.longrepr.reprcrash.message
INTERNALERROR> AttributeError: ReprFailDoctest instance has no attribute 'reprcrash'

=========================== 1 failed in 0.02 seconds ===========================
```

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/713

we have the same Issue with Python 3 and the current pytest version

```
» py.test sevenseconds/aws.py --doctest-modules
====================== test session starts ======================
platform linux -- Python 3.4.0 -- py-1.4.28 -- pytest-2.7.1
rootdir: /home/matthias/GIT/stups/sevenseconds, inifile: 
collected 1 items 

sevenseconds/aws.py F

=========================== FAILURES ============================
__________ [doctest] sevenseconds.aws.calculate_subnet __________
060     '''
061     >>> calculate_subnet(IPNetwork('10.0.0.0/16'), 'dmz', 0)
062     IPNetwork('10.0.0.0/21')
063 
064     >>> calculate_subnet(IPNetwork('10.0.0.0/16'), 'internal', 0)
065     IPNetwork('10.0.128.0/20')
066 
067     >>> calculate_subnet(IPNetwork('10.0.0.0/19'), 'dmz', 0)
Expected:
    IPNetwork('10.0.0.0/24')
Got:
    IPNetwork('10.0.0.0/21')

/home/matthias/GIT/stups/sevenseconds/sevenseconds/aws.py:67: DocTestFailure
=================== 1 failed in 0.28 seconds ====================
```


```
» py.test sevenseconds/aws.py --doctest-modules --junit-xml=junit.xml
====================== test session starts ======================
platform linux -- Python 3.4.0 -- py-1.4.28 -- pytest-2.7.1
rootdir: /home/matthias/GIT/stups/sevenseconds, inifile: 
collected 1 items 

sevenseconds/aws.py F
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/main.py", line 84, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/main.py", line 122, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/main.py", line 142, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 393, in execute
INTERNALERROR>     return wrapped_call(method(*args), self.execute)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 113, in wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 137, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 123, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/runner.py", line 65, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/runner.py", line 75, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/runner.py", line 123, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 521, in __call__
INTERNALERROR>     return self._docall(self.methods, kwargs)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 528, in _docall
INTERNALERROR>     firstresult=self.firstresult).execute()
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/core.py", line 394, in execute
INTERNALERROR>     res = method(*args)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/junitxml.py", line 180, in pytest_runtest_logreport
INTERNALERROR>     self.append_failure(report)
INTERNALERROR>   File "/usr/local/lib/python3.4/dist-packages/_pytest/junitxml.py", line 129, in append_failure
INTERNALERROR>     message = report.longrepr.reprcrash.message
INTERNALERROR> AttributeError: 'ReprFailDoctest' object has no attribute 'reprcrash'

=================== 1 failed in 0.25 seconds ====================
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -43,6 +43,7 @@ Mark Abramowitz
 Martijn Faassen
 Nicolas Delaby
 Piotr Banaszkiewicz
+Punyashloka Biswal
 Ralf Schmitt
 Ronny Pfannschmidt
 Ross Lawley

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,6 +1,9 @@
 2.8.0.dev (compared to 2.7.X)
 -----------------------------
 
+- fix issue713: JUnit XML reports for doctest failures.
+  Thanks Punyashloka Biswal.
+
 - Include setup and teardown in junitxml test durations.
   Thanks Janne Vanhala.
 

--- a/_pytest/junitxml.py
+++ b/_pytest/junitxml.py
@@ -123,10 +123,12 @@ class LogXML(object):
                 Junit.skipped(message="xfail-marked test passes unexpectedly"))
             self.skipped += 1
         else:
-            if isinstance(report.longrepr, (unicode, str)):
+            if hasattr(report.longrepr, "reprcrash"):
+                message = report.longrepr.reprcrash.message
+            elif isinstance(report.longrepr, (unicode, str)):
                 message = report.longrepr
             else:
-                message = report.longrepr.reprcrash.message
+                message = str(report.longrepr)
             message = bin_xml_escape(message)
             fail = Junit.failure(message=message)
             fail.append(bin_xml_escape(report.longrepr))

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -352,3 +352,16 @@ class TestDoctests:
         reprec = testdir.inline_run(p, "--doctest-modules",
                                     "--doctest-ignore-import-errors")
         reprec.assertoutcome(skipped=1, failed=1, passed=0)
+
+    def test_junit_report_for_doctest(self, testdir):
+        p = testdir.makepyfile("""
+            def foo():
+                '''
+                >>> 1 + 1
+                3
+                '''
+                pass
+        """)
+        reprec = testdir.inline_run(p, "--doctest-modules",
+                                    "--junit-xml=junit.xml")
+        reprec.assertoutcome(failed=1)

